src/vs/workbench/contrib/debug/browser/disassemblyView.ts

```diff
@@ -0,0 +1,242 @@
+			if (e.oldScrollTop > e.scrollTop && e.scrollTop < e.height) {
+				const topElement = Math.floor(e.scrollTop / this._fontInfo.lineHeight) + DisassemblyView.NUM_INSTRUCTIONS_TO_LOAD;
+				this.loadDisassembledInstructions(0, DisassemblyView.NUM_INSTRUCTIONS_TO_LOAD, '0x00004000');
+				this._disassembledInstructions!.reveal(topElement, 0);
+			} else if (e.oldScrollTop < e.scrollTop && e.scrollTop + e.height > e.scrollHeight - e.height) {
+				this.loadDisassembledInstructions(this._disassembledInstructions!.length, DisassemblyView.NUM_INSTRUCTIONS_TO_LOAD, '0x00004000');
+			}
+		});
+	}
+
+	layout(dimension: Dimension): void {
+		if (this._disassembledInstructions) {
+			this._disassembledInstructions.layout(dimension.height);
+		}
+	}
+
+	private loadDisassembledInstructions(index: number, numInstructions: number, address: string): void {
+		const newEntries: IDisassembledInstructionEntry[] = [];
+		for (let i = 0; i < numInstructions; i++) {
+			newEntries.push({ allowBreakpoint: true, isBreakpointSet: false, instruction: { address: `${address}${i}`, instruction: 'instruction instruction, instruction' } });
+		}
+		if (this._disassembledInstructions) {
+			// TODO: append/insert
+			this._disassembledInstructions.splice(index, 0, newEntries);
+		}
+	}
+
+}
+
+interface IBreakpointColumnTemplateData {
+	container: HTMLElement,
+	icon: HTMLElement
+}
+
+class BreakpointRenderer implements ITableRenderer<IDisassembledInstructionEntry, IBreakpointColumnTemplateData> {
+
+	static readonly TEMPLATE_ID = 'breakpoint';
+
+	templateId: string = BreakpointRenderer.TEMPLATE_ID;
+
+	renderTemplate(container: HTMLElement): IBreakpointColumnTemplateData {
+		const icon = append(container, $('.disassembly-view'));
+		icon.classList.add('codicon');
+		return { container, icon };
+	}
+
+	renderElement(element: IDisassembledInstructionEntry, index: number, templateData: IBreakpointColumnTemplateData, height: number | undefined): void {
+		// TODO: see getBreakpointMessageAndIcon in vs\workbench\contrib\debug\browser\breakpointEditorContribution.ts
+		//       for more types of breakpoint icons
+		if (element.allowBreakpoint) {
+			const breakpointIcon = 'codicon-' + icons.breakpoint.regular.id;
+			const breakpointHintIcon = 'codicon-' + icons.debugBreakpointHint.id;
+
+			if (element.isBreakpointSet) {
+				templateData.icon.classList.add(breakpointIcon);
+			}
+
+			templateData.container.onmouseover = () => {
+				templateData.icon.classList.add(breakpointHintIcon);
+			};
+
+			templateData.container.onmouseout = () => {
+				templateData.icon.classList.remove(breakpointHintIcon);
+			};
+
+			templateData.container.onclick = () => {
+				if (element.isBreakpointSet) {
+					element.isBreakpointSet = false;
+					templateData.icon.classList.remove(breakpointIcon);
+				} else if (element.allowBreakpoint) {
+					element.isBreakpointSet = true;
+					templateData.icon.classList.add(breakpointIcon);
+				}
+			};
+		}
```

Comment:
@isidorn I think there should be a better way to render breakpoints, as the breakpoint HTMLElement in main editor seems to allow duplicate items in classList. Do you know if I missed anything?

Reply:
Breakpoint in the main editor is rendered using editor glyph margins.
Unfortunately we do not have those for our custom editor, so we have to do something like this.

Just make sure to add all those listeners to a disposable list!

Reply:
They are not disposable so I can't register them in the disposable list. But I set them to null in the dispose function, and I think this should be fine?

Reply:
Please use `addStandardDisposableListener`
Code pointer https://github.com/microsoft/vscode/blob/main/src/vs/base/browser/dom.ts#L88